API v3 spec

bdaroz opened the pull request:
DO NOT MERGE

This is the, currently, work-in-progress on the API v3 spec. It contains the updated security model (API Key) as discussed in the meeting 10/29, as well as nearly all the endpoints as outlines in the "ideas" doc. (The `/team/x/event/x/status` and `/event/x/stats` endpoints are TBD until a model is defined (or spec'd from code).

The models are far from final, and encompass a few discussed items, but don't yet reflect the model definitions from the Ideas doc. Nor does it reflect the triage of API requests in the issues queue.... and many definitions/explanations are inadequate to downright missing.

So... much to do, but wanted to post a WIP in case someone else wants to jump in on it as well so work isn't duplicated.

bdaroz commented:
At this point all the models from v2 and the v3 idea doc are in place except an API status endpoint, and the 2016 score_breakdown model. After they get added I'll turn to the issue tracker for any API requests to see what can easily fit in the spec with the existing models or with minor tweaks to them.

bdaroz commented:
At this point the spec is fully fleshed out from the ideas doc, and some of the meeting discussion. It's rather large and somewhat unruly, but it can't be helped. (Swagger doesn't let you `$ref` response headers, which means `Last-Modified` and `Cache-Control` are duplicated for each path/endpoint.) HTML docs generated from this should be... interesting.

Two items pending before it's ready for review... I've bumped a handful of API-related issues and tagged #1507 in them to start a discussion on including those features in APIv3. Second, I have yet to do a pass to ensure parity between data available on the website and via the API.

phil-lopreiato commented:
At a quick skim, this seems good. A few changes though: 
 - Consolidate as much as you can (like the response headers and security blocks) into internal references.
 - Let's use json and not yaml. On the app-side, I've got some scripts that do minor mutations (by tweaking the json spec) before code generation, and I think that's valuable to be able to maintain

bdaroz commented:
I can flip it to JSON, but it'll still have the same problem -- you can't use internal references for response headers, only schemas, which bloats the file. I'll check about the security blocks, but I have a feeling it'll be the same problem.

phil-lopreiato commented:
[You sure about that](https://github.com/the-blue-alliance/the-blue-alliance-android/blob/master/libTba/swagger/apiv2-swagger.json#L535)? I think that you should be able to use references for any "sub-block" and I don't see a limitation to only schemas [in the docs](http://swagger.io/specification/#referenceObject)

bdaroz commented:
Ok, for responses... yes you can break them out to another section, however they're all generally unique (some of the `../keys` endpoints might have some schema commonality that's duplicated, but team keys vs event keys etc... and it's not *that* duplicative). The 304 is identical across all the paths, and yes, that can be broken out, and I'll do that, but it won't be saving any line count. :) (It's 1 line in the response, which gets replaced with a `$ref` tag and then 2-3 lines added for the referred item).

The docs only allow for a handful of top-level sections. `path`, `definitions` (schema), `parameters`, `responses`, `securityDefinitions` (that explain the headers/query values used, not that associate them to a `path`), `security` (which is used to define API-wide items, like these), and `tags`. (There are a few others, but nothing that covers "anything" in a sub-block).

It's unclear, and testing code gen concurs, that defining the parameters in `security` is supposed to be API-wide, but code gen doesn't appear to be enforcing them unless they are per-path.

Generally speaking, All the major models are `$ref`'ed, and the per-path schema covers one-offs and minor extensions (eg an array of Team objects). The major duplicative item is the `headers` block in each response. Even if I `$ref`'ed out the response, I'd still need to duplicate that as a header object isn't valid inside a `definitions` tag, for example.

I'll keep digging in case I missed something, but I don't think the response header subsection can be `$ref`'ed separate from the `response`.

bdaroz commented:
In re response headers: https://github.com/OAI/OpenAPI-Specification/issues/690

phil-lopreiato commented:
Huh, interesting, I stand corrected. I guess there isn't much more major to consolidate, then.

The only think I'm worried about is if we ever change/add some other header to every response, I don't want to have to write a script to codemode the spec. Maybe we don't need to include those in swagger and instead let clients handle it themselves/note it on a the docs webpage?

bdaroz commented:
I think the laudable goal is to be able to define a "contract", if you will, with the API spec, ideally in such a way as to be able to use it's self-documenting and code-generation functions, if not for TBA, for the clients.

If the spec is "tight" enough we get built in client libraries across a dozen or more languages. (Including ObjectiveC and Swift)

Unfortunately some of our calls are not concrete-enough for the spec (eg score_breakdown). There are ways around it, but it gets ugly (include the year in the call to the match endpoint) and probably not worth it -- but we should still define those models in our spec. (Regardless if it's Swagger or HTML or stone tablets.)

And yes, if we elect to add a new response header, there's going to be some serious regexp work on the spec file, but there's no way around it... yet.

Regardless of what we do with the spec file itself, or it's formatting, the bigger picture is to tack-down (no nails, yet) what schema changes / feature additions we want to include so work can start on the back end.

bdaroz commented:
Switched spec to JSON format. Will start walking the models for parity with the web pages later this week.

bdaroz commented:
I did a quick pass for parity between the information presented on the web pages and what's available via the API. Many of the v2 -> v3 changes were already reflected -- W-L-T counts, event stats, and so on. I did add a `home_championship` in the team model. To save the need to have to constantly alter the API if/when chaps moves from city to city it's spec'd as an `enum` with values of `north` and `south`.

I didn't catch anything else glaring missing... so this is probably ready for review and comment.

fangeugene commented:
Changes I'm making to the spec during implementation. WIP.
I'm open to discussion if anyone feels like the changes shouldn't be made.

Team stuff
- Full team model also includes `lat` and `lng`
- `state` renamed to `state_prov`
- `home_championship` looks like `{"2017": "St. Louis", "2018": "Detroit"}`
- `team/<team_key>/districts` is a dict like `{"2016": "2016fim", "2017": "2017fim"}`

bdaroz commented:
Two thoughts:

`home_championship` using the actual city names.... I'm torn between having a concrete spec (eg enum of `north` and `south`) vs a `String` returning an arbitrary city name value. With `north`/`south` You can know that team `x` attended the _same_ half-champs even if the actual city changes. On the other hand the client needs to learn/know that 2007 is St. Louis and 2018 is Detroit. Call it a toss-up for me, but pointing it out for discussion.

The return from `team/<team_key>/districts` does give me a tad bit of cringe from the DBA in me. My issue with returning `2016fim` in the result, as opposed to `fim`, is that the district is `fim` not `2016fim`. Put another way the `key` from the `districts/2016` call is `fim` not `2016fim`. Keys should be constant, in that the `team/<team_key>/districts` call and the `districts/<year>` call should return the same `key` for a district.

fangeugene commented:
Sure, we can do `fim` instead of `2016fim`. The change was more to make it a dict instead of a list. Both team robots and team districts return dict with the key being a year instead of a list.

gregmarra commented:
Is there a generic word for "state province prefecture whatever" that
people use for geo data?
On Tue, Jan 3, 2017 at 5:23 PM Eugene Fang <notifications@github.com> wrote:

> Sure, we can do fim instead of 2016fim. The change was more to make it a
> dict instead of a list. Both team robots and team districts return dict
> with the key being a year instead of a list.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/the-blue-alliance/the-blue-alliance/pull/1691#issuecomment-270272564>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADfDZlz0zZ-LQPNLaZ66DtqBCt9SMejks5rOvSDgaJpZM4KkpQK>
> .
>

fangeugene commented:
Google maps uses `administrative_area_level_1` Let's not.

bdaroz commented:
@gregmarra I don't think so... Google goes nuts with theirs, and Open Street Map, for example, takes all comers (some are `state` some `province`depending on the country). Since FRC is still predominantly US/English I think sticking with `state` should be fine... If you want to rename `city` to `locality` or something that would probably be fine (eg City vs Town vs Village vs Hamlet etc...), that seems more common in the mapping world.

bdaroz commented:
I'm updating the spec file, and have incorporated a number of changes from the v3 in testing. Some things I just haven't gotten to (Swagger and k/v pairs are a bit dicey), and some I have an issue with.

```
/team/{team_key}/events/keys
/team/{team_key}/events/{year}/keys
/team/{team_key}/event/{event_key}/matches/keys
/team/{team_key}/matches/{year}/keys
/events/{year}/keys
/event/{event_key}/teams/keys
/event/{event_key}/matches/keys
/district/{district_key}/{year}/events/keys
/district/{district_key}/{year}/teams/keys
```

These endpoints are all returning values like this right now:

```
[
  {
    "key": "frc1033"
  },
-- snip--
  {
    "key": "frc977"
  }
]
```

Do we really need to encapsulate the team key, in an object with a `key` property, in the array? Why just not an array of key strings?

```
/team/{team_key}/awards
Model: event_simple
```

Both of thee are still using internally-documented integer values as "types". Do we want to continue this forward in v3, or switch to API-defined enums?

```
/team/{team_key}/event/{event_key}/status
/district/{district_key}/{year}/rankings
```

Couldn't test, not implemented yet.

`Model: event_simple`

`event_district_key` is missing for non-district teams. Should probably return null. (Swagger and optional properties == ugly and a code gen nightmare, let's just not do it.)
`timezone` was spec'd for the simple model, only appearing on the full model.

`/event/{event_key}/alliances`

The missing, but spec'd `captain` property I think is a very useful piece of information. Sure, it's normally the 2nd listed team, but this makes it unambiguous to clients.

`/event/{event_key}/stats`

This is _very_ different from the spec. Spec was for game-specific stats, eg how many times a defence was breached in 2016. This is giving team-based OPR type stats. (Which we should include if we can....) How do we want to proceed with this call? One call covers what it does now, + game specific?

`/event/{event_key}/district_points`

This is also quite a bit different from the spec as well.

`/match/{match_key}/simple`

Recommend renaming `teams` to `team_keys` to clarify they are keys, not integers.

---
There may be more as I get into some things, and try to align some other stuff a bit, but at first pass this is what I see.

fangeugene commented:
Good feedback, I'll address them in a bit. In the meantime, here's the new ranking format https://www.thebluealliance.com/api/v3/event/2016casj/rankings

bdaroz commented:
Updated to reflect most of the changes in `/event/<event_key>/rankings` with one notable exception -- the spec continues to include both a quals W-L-T and an overall W-L-T, where implementation still only provides quals, and creates a disparity between HTML UI and API.

fangeugene commented:
Rankings will only have quals. the additional details will be in team/event/status.

fangeugene commented:
A few notes:
- We're probably going to keep enums as integers for everything. It's just too much of a pain to make parsable and human readable strings for everything, especially awards
- Alliances don't need "captain" because they're arranged in pick order, not driver station order.
- The same goes for Team_Event_status_alliance. `position` will be renamed to `pick`, with 0 being the first pick, 1 bein the second pick, etc, and -1 being a backup.

fangeugene commented:
Oh I forgot to add that matches now include surrogates.
Example: qm23 here https://www.thebluealliance.com/api/v3/event/2016casj/matches
```
"alliances": {
      "blue": {
        "score": 132, 
        "surrogates": [
          "frc254"
        ], 
        "teams": [
          "frc254", 
          "frc199", 
          "frc1700"
        ]
      }, 
      "red": {
        "score": 101, 
        "surrogates": [
          "frc5089"
        ], 
        "teams": [
          "frc5089", 
          "frc100", 
          "frc649"
        ]
      }
    }, 
```

bdaroz commented:
One thing about the alliance spec for matches -- it makes a point of using the property names `team_keys` and `surrogate_team_keys` to differentiate between `frc5089` and just `5089`. Do you want to update the code, or drop that clarification from the spec?

fangeugene commented:
I'll update property names to what you suggested.

fangeugene commented:
All of APIv3 is up except for team @ event status. All routes are here: https://github.com/the-blue-alliance/the-blue-alliance/blob/master/apiv3_main.py

bdaroz commented:
In, what could only be described as "you can't make this up", now as I get into testing in Java and C# codegen do we find that the `allOf` notation that lets us define a `Simple` schema and extend that into the full model, avoiding duplication, isn't supported by a number of the codegen languages. the remaining changes are minor, adding API function names, and aligning to actual results from the API.

bdaroz commented:
Current list of mismatches between spec and live API:

* `/status` does not have the `current_season_week` property.
* Team Model has the following fields included but API will always return `null`: `address`, `gmaps_place_id`, `gmaps_url`, `lat`, `lng`, `location_name`
* The district response model is still very different from the spec. Spec generally has the key for the district decoupled from the year (eg `mar` not `2017mar`). so `/team/{team}/districts` and `/districts/*` and `/district/*` are out of spec. The spec will be altered to reflect the API.
* `/event/{event}/stats not implemented
* `/event/{event}[oprs|insights|predictions] not in spec (may eliminate need for item above)
* `/event/{event}/district_points using key-value not arrays

runneals commented:
Is there a way to include the robot photo image / link + team media? This
would be extremely useful in scouting applications.

I believe in opendata, as such, we should make all of the data that is
shown accessible through an api. :)

On Wed, May 17, 2017 at 12:26 PM Brian Rozmierski <notifications@github.com>
wrote:

> Current list of mismatches between spec and live API:
>
>    - /status does not have the current_season_week property.
>    - Team Model has the following fields included but API will always
>    return null: address, gmaps_place_id, gmaps_url, lat, lng,
>    location_name
>    - The district response model is still very different from the spec.
>    Spec generally has the key for the district decoupled from the year (eg
>    mar not 2017mar). so /team/{team}/districts and /districts/* and
>    /district/* are out of spec. The spec will be altered to reflect the
>    API.
>    - `/event/{event}/stats not implemented
>    - `/event/{event}[oprs|insights|predictions] not in spec (may
>    eliminate need for item above)
>    - `/event/{event}/district_points using key-value not arrays
>
> —
> You are receiving this because you are subscribed to this thread.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/the-blue-alliance/the-blue-alliance/pull/1691#issuecomment-302165035>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AByuQaUNlze_YCpZDDFE0qL2jN1kBayAks5r6y3bgaJpZM4KkpQK>
> .
>
-- 
Sent from my iPhone

bdaroz commented:
Like this?

https://www.thebluealliance.com/api/v3/team/frc20/media/2017

runneals commented:
Spot on :)

On Wed, May 17, 2017 at 12:59 PM, Brian Rozmierski <notifications@github.com
> wrote:

> Like this?
>
> https://www.thebluealliance.com/api/v3/team/frc20/media/2017
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/the-blue-alliance/the-blue-alliance/pull/1691#issuecomment-302175183>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AByuQWkPlf06DE5ju-pWT0oMkZQMvgIWks5r6zWagaJpZM4KkpQK>
> .
>

bdaroz commented:
Two items left to fix:

* `/status` does not have the `current_season_week` property.
* `/event/{event}/district_points` using key-value not arrays

(I thought I fixed the last one but the quick test of codegen pointed it out, next commit will fix it.)

bdaroz commented:
Ok, updated my branch from current `master` in TBA repo. Also renamed the spec file to match the trusted one and updated the main controller and frame renderer to point to this spec file instead of the one in the Android repo.
